frame/contracts/src/exec.rs

```diff
@@ -378,456 +396,640 @@ pub trait Executable<T: Config>: Sized {
 				// with the exception of a contract that called `seal_terminate`.
-				Rent::<T, E>::charge(&dest, contract, occupied_storage)?
-					.ok_or(Error::<T>::NewContractNotFunded)?;
+				let contract = Rent::<T, E>
+					::charge(&account_id, frame.invalidate(), occupied_storage)
+					.map_err(|e| (e.into(), code_len))?
+					.ok_or((Error::<T>::NewContractNotFunded.into(), code_len))?;
+				frame.contract_info = CachedContract::Cached(contract);
 
 				// Deposit an instantiation event.
-				deposit_event::<T>(vec![], Event::Instantiated(caller.clone(), dest.clone()));
+				deposit_event::<T>(vec![], Event::Instantiated(
+					self.caller().clone(),
+					account_id,
+				));
+			}
 
-				Ok(output)
-			});
+			Ok((output, code_len))
+		};
 
-			use frame_support::storage::TransactionOutcome::*;
+		// All changes performed by the contract are executed under a storage transaction.
+		// This allows for roll back on error. Changes to the cached contract_info are
+		// comitted or rolled back when popping the frame.
+		let (success, output) = with_transaction(|| {
+			let output = do_transaction();
 			match output {
-				Ok(_) => Commit(output),
-				Err(_) => Rollback(output),
+				Ok((ref result, _)) if result.is_success() => {
+					TransactionOutcome::Commit((true, output))
+				},
+				_ => TransactionOutcome::Rollback((false, output)),
 			}
-		})?;
+		});
+		self.pop_frame(success);
+		output
+	}
 
-		Ok((dest, output))
+	/// Remove the current (top) frame from the stack.
+	///
+	/// This is called after running the current frame. It commits cached values to storage
+	/// and invalidates all stale references to it that might exist further down the call stack.
+	fn pop_frame(&mut self, persist: bool) {
+		// Revert the account counter in case of a failed instantiation.
+		if !persist && self.frame().entry_point == ExportedFunction::Constructor {
+			self.account_counter.as_mut().map(|c| *c = c.wrapping_sub(1));
+		}
+
+		// Pop the current frame from the stack and return it in case it needs to interact
+		// with duplicates that might exist on the stack.
+		let frame = self.frames.pop();
+
+		if let Some(frame) = frame {
+			let prev = self.frame_mut();
+			let account_id = &frame.account_id;
+			prev.nested_meter.absorb_nested(frame.nested_meter);
+			// Only gas counter changes are persisted in case of a failure.
+			if !persist {
+				return;
+			}
+			if let CachedContract::Cached(contract) = frame.contract_info {
+				// optimization: Predecessor is the same contract.
+				// We can just merge into the predecesor without a storage write.
+				if prev.account_id == *account_id {
+					prev.contract_info = CachedContract::Cached(contract);
+					return;
+				}
+
+				// Predecessor is a different contract: Write to storage and invalidate all
+				// references to it.
+				<ContractInfoOf<T>>::insert(account_id, ContractInfo::Alive(contract));
+				if let Some(c) = self.frames_mut().skip(1).find(|f| f.account_id == *account_id) {
```

Comment:
Why do we skip the first frame ? can't it be invalidated ?

Reply:
The `frames_*()` iterator goes in reverse direction. The first element is the top of the stack. If the top of the stack is the same contract as the popped one we don't invalidate but merge (see the optimization above this fragment).

Reply:
ah I missed the reversed iteration part, all good then